### ISSUE TITLE
trial twisted/mail/ traces back, omitting trailing / does not.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2005]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)| @jerub reported|
|-|-|
|Trac ID|trac#2005|
|Type|defect|
|Created|2006-08-19 12:57:34Z|

```
$ trial twisted/mail/
Traceback (most recent call last):
  File "/Users/stephen/src/Twisted/trunk/bin/trial", line 24, in ?
    run()
  File "/Users/stephen/src/Twisted/trunk/twisted/scripts/trial.py", line 346, in run
    suite = _getSuite(config)
  File "/Users/stephen/src/Twisted/trunk/twisted/scripts/trial.py", line 308, in _getSuite
    return loader.loadByNames(config['tests'], recurse)
  File "/Users/stephen/src/Twisted/trunk/twisted/trial/runner.py", line 429, in loadByNames
    suites = [self.loadAnything(thing, recurse)
  File "/Users/stephen/src/Twisted/trunk/twisted/trial/runner.py", line 398, in loadAnything
    return self.loadPackage(thing, recurse)
  File "/Users/stephen/src/Twisted/trunk/twisted/trial/runner.py", line 364, in loadPackage
    pkgobj = modules.getModule(package.__name__)
  File "/Users/stephen/src/Twisted/trunk/twisted/python/modules.py", line 625, in getModule
    return theSystemPath[moduleName]
  File "/Users/stephen/src/Twisted/trunk/twisted/python/modules.py", line 569, in __getitem__
    pe = PathEntry(
  File "/Users/stephen/src/Twisted/trunk/twisted/python/modules.py", line 525, in _findEntryPathString
    assert rval in self.importerCache, '%r for %r not in import cache %s' % (
AssertionError: 'twisted' for <module '__init__' from 'twisted/mail/__init__.pyc'> not in import cache {'/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4': None,
 '/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/': None,
 '/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/encodings': None,
 '/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/lib-dynload': None,

```

<details><summary>Searchable metadata</summary>

```
trac-id__2005 2005
type__defect defect
reporter__jerub jerub
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1155992254000000 1155992254000000
changetime__1173692841000000 1173692841000000
version__None None
owner__ 
cc__spiv cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

See also [#2489](https://github.com/twisted/twisted/issues/2489) and [#2507](https://github.com/twisted/twisted/issues/2507)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

Maybe comes from this:

```
>>> from twisted.python import reflect
>>> reflect.filenameToModuleName('twisted/test')
'twisted.test'
>>> reflect.filenameToModuleName('twisted/test/')
'twisted.'
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|spiv commented|
|-|-|

This is really irritating and confusing, and clearly a regression for earlier versions.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

Ready to review in trial-slash-2005.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_7"></a>|@radix set owner to @therve|
|-|-|

It'd be niceish if the code that says "if not base:" had a small comment like "# this happens when fn ends with a path separator".

I'm glad you tested it at the trial level too. Thanks for this, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_8"></a>|@radix commented|
|-|-|

Oh right, one more thing: please just "import os" instead of "os.path".

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set status to `closed`|
|-|-|

(In [19823]) Merge trial-slash-2005

Author: therve
Reviewer: radix
Fixes [#2005](https://github.com/twisted/twisted/issues/2005)

Workaround a problem with reflect.filenameToModuleName which cause unexpected
behavior with trial: 'trial foo/bar' worked whereas 'trial foo/bar/' didn't.
Also, add tests for trial's loader and for reflect.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
